src/Pipeline.ts

```diff
@@ -0,0 +1,108 @@
+import Commander from 'ioredis/built/utils/Commander';
+import {Command} from 'ioredis';
+import {MomentoRedisAdapter} from './momento-redis-adapter';
+import {ArgumentType} from 'ioredis/built/Command';
+import {CommandParameter} from 'ioredis/built/types';
+
+interface Pipeline {
+  length: number;
+}
+
+class Pipeline extends Commander<{type: 'pipeline'}> {
+  promise: Promise<[error: Error | null, result: unknown][] | null>;
+  resolve: (result: [error: Error | null, result: unknown][] | null) => void;
+  reject: (error: Error) => void;
+
+  private _queue: Array<Command> = [];
+  private _result: Array<unknown> = [];
+
+  momentoAdapter: MomentoRedisAdapter;
+
+  constructor(public redis: MomentoRedisAdapter) {
+    super();
+    this.momentoAdapter = redis;
+
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const _this = this;
+    Object.defineProperty(this, 'length', {
+      get: function () {
+        return _this._queue.length;
+      },
+    });
+  }
+
+  sendCommand(command: Command): unknown {
+    command.pipelineIndex = this._queue.length;
+    command.promise = this.invokeMomentoRedisClient(command.name, command.args);
+    this._queue.push(command);
+    return this;
+  }
+
+  addBatch(commands: Array<Array<unknown>>) {
+    for (let i = 0; i < commands.length; ++i) {
+      // Parse input cmd in Array format ex: ["get", "foo"]
+      const command = commands[i];
+      const commandName = command[0] as string;
+      const args = command.slice(1) as ArgumentType[];
+
+      // Invoke Command
+      const cmdToQueue = new Command(commandName, args);
+      cmdToQueue.promise = this.invokeMomentoRedisClient(
+        cmdToQueue.name,
+        cmdToQueue.args
+      );
+
+      // Push command with promise to queue
+      cmdToQueue.pipelineIndex = this._queue.length;
+      this._queue.push(cmdToQueue);
+    }
+    return this;
+  }
+
+  // Invoke function on momento adaptor
+  invokeMomentoRedisClient(
+    name: string,
+    args: CommandParameter[]
+  ): Promise<unknown> {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    if (this.momentoAdapter[name] === undefined) {
+      throw new Error(`Un-Supported Command Passed: ${name}`);
+    }
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+    return this.momentoAdapter[name](...args); // eslint-disable-line @typescript-eslint/no-unsafe-return
+  }
+
+  exec(): Promise<[error: Error | null, result: unknown][] | null> {
+    if (!this._queue.length) {
+      this.resolve([]);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const _this = this;
```

Comment:
Why do you need to do this?

Reply:
So we can access outer scope `this` from within the `execPipeline` nested function. Cant use `this` directly
<img width="847" alt="Screenshot 2024-04-26 at 8 26 38 AM" src="https://github.com/momentohq/momento-node-ioredis-client/assets/5491827/a917349c-3263-4b7e-8480-8dab0855076b">

Is not the greatest its bit of dirty JS TBH. Is how IORedis was doing as well why started here. I can maybe try come back and refactor in cleaner way. Maybe in follow up pr though?